service/history/historyEngine.go

```diff
@@ -2700,30 +2717,66 @@ func (e *historyEngineImpl) applyWorkflowIDReusePolicyForSigWithStart(
 	prevCloseState := prevExecutionInfo.CloseStatus
 
 	return e.applyWorkflowIDReusePolicyHelper(
+		ctx,
 		prevStartRequestID,
 		prevRunID,
 		prevState,
 		prevCloseState,
 		execution,
 		wfIDReusePolicy,
+		domainID,
 	)
+}
+
+func (e *historyEngineImpl) requestTerminateWorkflow(
+	ctx context.Context,
+	domainID string,
+	execution workflow.WorkflowExecution,
+	targetRunID string,
+) error {
 
+	termRequest := &h.TerminateWorkflowExecutionRequest{
+		DomainUUID: common.StringPtr(domainID),
+		TerminateRequest: &workflow.TerminateWorkflowExecutionRequest{
+			WorkflowExecution: &workflow.WorkflowExecution{
+				WorkflowId: execution.WorkflowId,
+				RunId:      common.StringPtr(targetRunID),
+			},
+			Reason:   common.StringPtr(TerminateIfRunningReason),
+			Details:  []byte(fmt.Sprintf(TerminateIfRunningDetailsTemplate, execution.GetRunId())),
+			Identity: common.StringPtr(TerminateIfRunningIdentity),
+		},
+	}
+	err := e.TerminateWorkflowExecution(ctx, termRequest)
+	if err != nil && err != ErrWorkflowCompleted {
+		return err
+	}
+	return nil
 }
 
 func (e *historyEngineImpl) applyWorkflowIDReusePolicyHelper(
+	ctx context.Context,
 	prevStartRequestID,
 	prevRunID string,
 	prevState int,
 	prevCloseState int,
 	execution workflow.WorkflowExecution,
 	wfIDReusePolicy workflow.WorkflowIdReusePolicy,
+	domainID string,
 ) error {
 
 	// here we know some information about the prev workflow, i.e. either running right now
 	// or has history check if the workflow is finished
 	switch prevState {
 	case persistence.WorkflowStateCreated,
 		persistence.WorkflowStateRunning:
+		if wfIDReusePolicy == workflow.WorkflowIdReusePolicyTerminateIfRunning {
+			err := e.requestTerminateWorkflow(ctx, domainID, execution, prevRunID)
```

Comment:
I think I may missed something here. How do we ensure terminate and start are done in one transaction?

Reply:
The problem to solve is: when user terminate workflow then start, they may see WorkflowAlreadyExist. 
With this change, the `requestTerminateWorkflow` is called with current execution lock so others are block to start. Then it terminate the current run, and start a new one in one "semantic" transaction (not a db transaction).  

This part is indeed not-straightforward and may confuse people, let me know if that make sense

Reply:
I will also add these comments to code

Reply:
Two questions:
1. If the lock is acquired in the start workflow helper, then when it calls terminate, how can terminate get the lock?
2. For any reason, if the call stops after it terminates the workflow. Then could user see the behavior like call startWorkflowExecution and received a context timeout then the current open workflow gets terminated?

Reply:
Synced offline, but the short answers are:
1. the lock are different. lock in startworkflowhelper is for current execution, terminate is using concrete execution lock
2. yes, user will see current get terminated but start failed.

The 2nd question is super and thanks for bring it up. Let me think more on best option.

Reply:
I will do rewrite this and make it in one db transaction